File: Java/Exemple/src/exercice_objet_bouteille/bouteille.java
package exercice_objet_bouteille;

public class bouteille {
		
		//attributs
		private final double capaciteEnL;
		private double contenanceEnL;
		private boolean estOuverte;
		private String nom;
		
		//constructeur par defaut
		public bouteille()
		{
			nom = " ";
			contenanceEnL = 0.9;
			capaciteEnL = 1;
			estOuverte = true;
		}
		
		//construteur avec parametre
		public bouteille(String _nom, double _contenanceEnL, double _capaciteEnL, boolean _estOuverte)
		{
			this.nom = _nom;
			this.contenanceEnL = _contenanceEnL;
			this.capaciteEnL = _capaciteEnL;
			this.estOuverte = _estOuverte;
		}
		
		public boolean ouvrir()
		{
			if(!this.estOuverte)
			{
				this.estOuverte = true;
				return true;
			}
			else
			{
				return false;
			}
			
		}
		
//		public void ouvrirProcedure() //procedure
//		{
//			if(this.estOuverte)
//			{
//				this.estOuverte = true;
//			}
//		}
		
		public boolean fermer()  //fonction
		{
			if(this.estOuverte)
			{
				this.estOuverte = false;
				return true;
			}
			else
			{
				return false;
			}
		}
		
		/*public boolean viderEntierement()
		{
			if(this.estOuverte)
			{
				this.contenanceEnL = 0;
				return true;
			}
			else
			{
				return false;
			}
		}
		
		public boolean remplirEntierement()
		{
			if(this.estOuverte)
			{
				this.contenanceEnL = this.capaciteEnL;
				return true;
			}
			else
			{
				return false;
			}
			
		}
		
		public boolean completerEnL()
		{
			if(this.contenanceEnL < this.capaciteEnL)
			{
				this.contenanceEnL = this.capaciteEnL;
				return true;
			}
			else
			{
				return false;
			}
		{*/
			
		public boolean remplirTout()
		{
			if(this.estOuverte)
			{
				if(this.contenanceEnL < this.capaciteEnL)
				{
					this.contenanceEnL = this.capaciteEnL;
					return true;
				}
				else
				{
					return false;
				}
			}
			else
			{
				return false;
			}
			
		}
		
		public boolean viderTout()
		{
			if(this.estOuverte)
			{
				if(this.contenanceEnL > 0)
				{
					this.contenanceEnL = 0;
					return true;
				}
				else
				{
					return false;
				}
			}
			else
			{
				return false;
			}
		}
		
		public boolean remplir(double quantiteEnL)
		{
			double plein = quantiteEnL + this.contenanceEnL;
			if(this.estOuverte)
			{
				if(plein <= this.capaciteEnL)
				{
					this.contenanceEnL = plein;
					return true;
				}
				else
				{
					return false;
				}
			}
			else
			{
				return false;
			}
		}
		
		public boolean vider(double quantiteEnL)
		{
			if(this.estOuverte)
			{
				if(quantiteEnL > 0)
				{
					if(this.contenanceEnL - quantiteEnL >= 0)
					{
						this.contenanceEnL = this.contenanceEnL - quantiteEnL;
						return true;
					}
					else
					{
						return false;
					}
				}
				else
				{
					return false;
				}
			}
			else
			{
				return false;
			}
		}
		
}

File: C#/SolutionBouteille/ClassLibraryBouteille/Bouteille.cs
using System.ComponentModel.Design;

namespace ClassLibraryBouteille
{
    public class Bouteille
    {
        //Attributs
        private double capaciteMaxEnMl;
        private bool estOuverte;
        private double quantiteLiquideEnMl;
        //private DateTime dateDePeremption;

        //SET/GET
        public double CapaciteMaxEnMl { get => capaciteMaxEnMl; set => capaciteMaxEnMl = value; }
        public bool EstOuverte { get => estOuverte; set => estOuverte = value; }
        public double QuantiteLiquideEnMl { get => quantiteLiquideEnMl; set => quantiteLiquideEnMl = value; }

        



        //Constructeurs par defaut
        public Bouteille()
        {
            this.capaciteMaxEnMl = 1000;
            this.estOuverte = false;
            this.quantiteLiquideEnMl = 1000;
        }

        //Constructeur avec semi-parametre
        public Bouteille(double _capaciteMaxEnMl)
        {
            this.capaciteMaxEnMl = _capaciteMaxEnMl;
            this.estOuverte = false;
            this.quantiteLiquideEnMl = _capaciteMaxEnMl;
        }

        /*Constructeur avec parametre
        public Bouteille(double _capaciteMaxEnMl, bool _estOuverte, double _quantiteLiquideEnMl) /*DateTime _dateDePeremption)
        {
            this.capaciteMaxEnMl = _capaciteMaxEnMl;
            this.estOuverte = _estOuverte;
            this.quantiteLiquideEnMl = _quantiteLiquideEnMl;
            this.dateDePeremption = _dateDePeremption;
        }*/

        //Methodes
        public bool Ouvrir()
        {
            if(!this.estOuverte)
            {
                this.estOuverte=true;
                return true;
            }
            else
            {
                return false;
            }     
        }

        public bool Fermer()
        {
            if(this.estOuverte)
            {
                this.estOuverte = false;
                return true;
            }
            else
            {
                return false;
            }
        }
        
        public bool ViderTout()
        {
            if(this.estOuverte && this.quantiteLiquideEnMl > 0)
            {
                this.quantiteLiquideEnMl = 0;
                return true;
            }
            else
            {
                return false;
            }
        }

        public bool RemplirTout()
        {
            if(this.estOuverte && this.quantiteLiquideEnMl < this.capaciteMaxEnMl)
            {
                this.quantiteLiquideEnMl = this.capaciteMaxEnMl;
                return true;
            }
            else
            {
                return false;
            }
        }

        public bool Remplir(double _quantiteLiquideEnMl) //_quantiteLiquideEnMl c'est ce qu'on ajoute dans la bouteille
        {
            if(this.estOuverte && this.quantiteLiquideEnMl < this.capaciteMaxEnMl)
            {
                if(this.quantiteLiquideEnMl + _quantiteLiquideEnMl > this.capaciteMaxEnMl)
                {
                    this.quantiteLiquideEnMl = this.capaciteMaxEnMl;
                }
                else
                {
                    this.quantiteLiquideEnMl += _quantiteLiquideEnMl;
                }
                return true;
            }
            else
            {
                return false;
            }
        }

        public bool Vider(double _quantiteLiquideEnMl) // _quantiteLiquideEnMl c'est ce qu'on retire de la bouteille
        {
            if(this.estOuverte && this.quantiteLiquideEnMl> 0)
            {
                if(this.quantiteLiquideEnMl - _quantiteLiquideEnMl <= 0)
                {
                    this.quantiteLiquideEnMl = 0;
                }
                else
                {
                    this.quantiteLiquideEnMl -= _quantiteLiquideEnMl;
                }
                return true;
            }
            else
            {
                return false;
            }
        }

    }
}